### PR TITLE
Feature/distractor analysis ui

### PR DESCRIPTION
**Pull request by bgsavage**
Apologies in advanced for the large PR.  I'll highlight areas of interest.

Note:  I'm assuming there will be a task in the API implementation to circle back and update all my //TODOs and wire up the ui to the api result.  Should be pretty straight forward.  I'll also be updating the API requirements with what the ui is expecting.

image

image

### CODE REVIEW

File: webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts [hunk 1]
```diff
@@ -48,6 +48,38 @@ export class AssessmentExamMapper {
         assessmentItem.scores.push(this.mapExamItemFromApi(apiExamItem));
       }
 
+      // TODO: DWR-1068: Move all of the below mapping to mapAssessmentItemFromApi method.
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/assessment-exam.mapper.ts]: I'm anticipating we'll just be updating the existing assessment items api endpoint with the new data points we are interested in.  Once we do so, it should just be a straight mapping that we set in `mapAssessmentItemFromApi` and the below code I temporarily added can be removed.

File: webapp/src/main/webapp/src/app/assessments/assessments.component.html [hunk 2]
```diff
@@ -123,7 +123,6 @@ <h2 class="blue-dark h3 mt-lg mb-md">
                       [filterBy]="clientFilterBy"
                       [allowFilterBySessions]="allowFilterBySessions"
                       [assessmentProvider]="assessmentProvider"
-                      [loadAssessmentItems]="boundLoadAssessmentItems"
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/assessments.component.html]: unrelated cleanup of unused param.

File: webapp/src/main/webapp/src/app/assessments/assessments.component.ts [hunk 3]
```diff
@@ -231,10 +227,6 @@ export class AssessmentsComponent implements OnInit {
     this.filterOptions.hasSummative = this.selectedAssessments.some(a => a.isSummative);
   }
 
-  private loadAssessmentItems(assessmentId: number): Observable<AssessmentItem[]> {
-    return this.assessmentProvider.getAssessmentItems(assessmentId);
-  }
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/assessments.component.ts]: unrelated cleanup

File: webapp/src/main/webapp/src/app/assessments/items/item-scores/item-scores.component.html [hunk 4]
```diff
@@ -16,6 +16,7 @@
       {{ score.enrolledGrade | gradeDisplay:'enrolled-name' }}
     </ng-template>
   </p-column>
+  <p-column *ngIf="includeResponse" field="response" header="{{'labels.assessments.items.tabs.scores.response' | translate}}" [sortable]="true"></p-column>
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/items/item-scores/item-scores.component.html]: I added an optional response column to the student table, thinking this would be useful information when looking at the student scores in the context of distractor analysis.  (See second screen shot in the description of this PR)

File: webapp/src/main/webapp/src/app/assessments/model/assessment-item.model.ts [hunk 5]
```diff
@@ -13,7 +13,11 @@ export class AssessmentItem {
   difficulty: string;
   commonCoreStandardIds: string[];
   maxPoints: number;
+  numberOfChoices: number;
   scores: ExamItemScore[] = [];
+  isMultipleChoice: boolean;
+  isMultipleSelect: boolean;
+  answerKey: string;
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/model/assessment-item.model.ts]: Here you can see at a glance the new fields I'm expecting the api to return for the existing assessment items call.

File: webapp/src/main/webapp/src/app/assessments/model/item-point-field.model.ts [hunk 6]
```diff
@@ -1,9 +1,9 @@
 /**
- * This model represents a "Points" column in the Items by Points Earned
+ * This model represents a dynamic column in the Items by Points Earned or DistractorAnalysis
  * display table and export.
  */
-export class ItemPointField {
+export class DynamicItemField {
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/model/item-point-field.model.ts]: Renamed for reuse with the Distractor Analysis which also has similar dynamic columns.

File: webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts [hunk 7]
```diff
@@ -2,14 +2,16 @@ import { Injectable } from "@angular/core";
...
-import { ItemPointField } from "../model/item-point-field.model";
+import { DynamicItemField } from "../model/item-point-field.model";
 import * as math from "mathjs";
 
 @Injectable()
 export class ExamStatisticsCalculator {
   private readonly NumberFieldPrefix = "number-point_";
   private readonly PercentFieldPrefix = "percent-point_";
 
+  private readonly potentialResponses = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts]: @jtreuting and I tossed around the idea of using a hardcoded string of potential responses or just ascii coding it.  We thought this would be readable, and would more easily accommodate for change should the potentialResponses change in the future.

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.html [hunk 8]
```diff
@@ -1,3 +1,81 @@
...
+</div>
+
+<div *ngIf="loading" class="loader">
+  <div class="inner">
+    <i class="fa fa-spinner fa-pulse fa-3x fa-fw"></i><span class="text">{{ 'messages.loading' | translate }}</span>
+  </div>
+</div>
+<p-dataTable *ngIf="!loading" #datatable [value]="filteredMultipleChoiceItems" sortField="position"
+             tableStyleClass="table table-striped table-hover" [tableStyle]="{'table-layout':'auto'}"
+             expandableRows="true" rowExpandMode="single">
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.html]: This table will look very familiar to the Items By Points Earned table as there were a lot of similarities.  There are subtle differences though, and had the thought to create an item-table component that both results-by-item and distractor-analysis use.

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts [hunk 9]
```diff
@@ -1,14 +1,109 @@
...
 
+  // Unfortunately, this is a bit of dom hijacking to set the parent <td> class to green
+  // since primeng datatable does not currently support a setCellStyle function.
+  // https://github.com/primefaces/primeng/issues/2157
+  setTdClass(cell, item: AssessmentItem, column: DynamicItemField) {
+    if(item.answerKey && item.answerKey.indexOf(column.label)!==-1) {
+      let td = cell.parentNode.parentNode;
+      this.renderer.addClass(td, "green");
+    }
+  }
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: Spent a couple hours longer than anticipated today just trying to get the cells green.  This was the solution I ended up using as a suggested workaround for the referenced feature request.

File: webapp/src/main/webapp/src/app/assessments/model/assessment-item.model.ts (hunk 5 above)
Comment (jtreuting) [reply webapp/src/main/webapp/src/app/assessments/model/assessment-item.model.ts]: Shouldn't the API just return the item type instead and the front end can decide what to do with it.

Comment (bgsavage) [reply webapp/src/main/webapp/src/app/assessments/model/assessment-item.model.ts]: Good point, these two bools should evolve into readonly helper getters using an underlying type of `itemType` to return the result.

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts [hunk 10]
```diff
@@ -1,14 +1,109 @@
...
+  private _exams: Exam[];
+
+  constructor(private examCalculator: ExamStatisticsCalculator, private renderer: Renderer2) {
+  }
 
   ngOnInit() {
+    this.loading = true;
+    this.assessmentProvider.getAssessmentItems(this.assessment.id).subscribe(assessmentItems => {
+      // TODO: DWR-1068: Allow this filter to be applied as a query param to the api.
+      let multipleChoiceItems = assessmentItems.filter(x => x.isMultipleChoice || x.isMultipleSelect);
```
Comment (bgsavage) [inline webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: Ideally the API takes a query param of item types to search for so that it doesn't return item data that we throw out right here.

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts (hunk 9 above)
Comment (jtreuting) [reply webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: Just thinking how this is tied to their implementation to trust that 2 steps up is the TD.  Not sure if it's cleaner or just added complexity to step up until you find the TD explicitly?  Not sure

File: webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts [hunk 11]
```diff
@@ -41,6 +41,12 @@ export class ItemTabComponent implements OnInit {
   @Input()
   showStudentScores: boolean = true;
 
+  /**
+   * If true, adds a column to the student scores which is the student's response.
+   */
+  @Input()
+  includeResponseInStudentScores: boolean = false;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts]: should we call this var showResponse*? I notice we typically use show* when naming show/hide flags

File: webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts [hunk 12]
```diff
@@ -61,33 +63,82 @@ export class ExamStatisticsCalculator {
...
+      for (let i = 0; i < item.numberOfChoices; i++) {
+        let response = this.potentialResponses[ i ];
+
+        if (item.scores.length > 0) {
+          let compareFunction = item.isMultipleSelect
+            ? (x => x.response != null && x.response.indexOf(response) !== -1)
+            : (x => x.response == response);
+
+          let count = item.scores.filter(compareFunction).length;
+          item[ this.NumberFieldPrefix + response ] = count;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts]: Should we add comments that these methods modify the input params? I would want to know this when calling the function.

File: webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts [hunk 13]
```diff
@@ -61,33 +63,82 @@ export class ExamStatisticsCalculator {
...
+      column.label = response;
       pointFields.push(column);
     }
 
     return pointFields;
   }
+
+  assertNumberOfChoicesIsValid(numberOfChoices) {
+    // Assert we have a defined potential response for the given number of choices.
+    if (numberOfChoices >= this.potentialResponses.length) {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts]: so when we get the item info it doesn't tell us the available options the student had for the item i'm guessing?

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.html [hunk 14]
```diff
@@ -1,3 +1,81 @@
...
+  <div class="z-index-over-data-table">
+    <button class="btn btn-default btn-xs z-index-over-data-table"
+            (click)="exportDistractorAnalysis()">
+      <i class="fa fa-bold fa-table"></i>
+      {{'labels.export.distractor-analysis' | translate}}
+    </button>
+  </div>
+</div>
+
+<div *ngIf="loading" class="loader">
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.html]: We should probably make this snippet a common comp at some point i think we have 3 or 4 scattered around with the same markup

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts [hunk 15]
```diff
@@ -1,14 +1,109 @@
...
+  getChoiceRowStyleClass(index: number) {
+    return index == 0
+      ? 'level-down'
+      : '';
   }
 
+  // Unfortunately, this is a bit of dom hijacking to set the parent <td> class to green
+  // since primeng datatable does not currently support a setCellStyle function.
+  // https://github.com/primefaces/primeng/issues/2157
+  setTdClass(cell, item: AssessmentItem, column: DynamicItemField) {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: is it possible to define a div in a row template and style that instead? might end up being less hacky

Comment (bgsavage) [reply webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: It won't look as nice as the bg color will only be a portion of the entire td.  I tried that first. :)

File: webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts (hunk 13 above)
Comment (bgsavage) [reply webapp/src/main/webapp/src/app/assessments/results/exam-statistics-calculator.ts]: API needs to be updated to include that info.  Regardless, I still want to make this check so we don't end up with a random index out of range error further down in the code.

File: webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts (hunk 11 above)
Comment (bgsavage) [reply webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts]: We actually already have a 'showResponse' for the tab itself.

File: webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts (hunk 15 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/view/distractor-analysis/distractor-analysis.component.ts]: i see. you'd have to apply a style to the table and mess with specific cell padding and junk too, which is also nasty. it might be preferable to do it that way though if we have time to revisit.

File: webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts (hunk 11 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/items/item-tab.component.ts]: gotcha. showResponseColumn? showResponse2? x P
